### PR TITLE
Support per-channel quantized INT8 weights unpacking in XNNPACK delegate

### PR DESCRIPTION
**Pull request by dev0x13**
This MR expands INT8 weights unpacking for FP32 inference in XNNPACK delegate support added in [the previous MR](https://github.com/tensorflow/tensorflow/pull/48751) for per-channel dynamic range quantized model. Previous changes have only supported per-tensor quantization mode, which is obsolete in recent TensorFlow releases.
I have not added proper testing yet, because I'd like to get some suggestions from a maintainer, i.e. the best way to organize such testing while keeping the codebase clean (specifically, how to change testers). Thank you in advance!

### CODE REVIEW

**Comment by dev0x13** [comment]
@multiverse-tf Could you review this please?

**Comment by gbaned** [comment]
@dev0x13  Can you please resolve conflicts? Thanks!

**Comment by dev0x13** [comment]
@gbaned Done

**Comment by dev0x13** [comment]
@multiverse-tf Could you review this please?

**Comment by dev0x13** [comment]
It's been more than a month since I've created this MR. Should I close it now due to the lack of review?

**Comment by multiverse-tf** [comment]
> It's been more than a month since I've created this MR. Should I close it now due to the lack of review?

Sorry for not catching this review request earlier. In the past month, I think we've also added native per-channel quantized INT8 op support in XNNPACK. @Maratyszcza, could you shed more light on the support? Thx

**Comment by dev0x13** [comment]
Closing due to the lack of response from maintainers.

**Comment by multiverse-tf** [comment]
@Maratyszcza could you help with this review? Thx!

**Comment by dev0x13** [comment]
@multiverse-tf @Maratyszcza Could you review this please? Thanks in advance!

**Comment by dev0x13** [comment]
@multiverse-tf @Maratyszcza Could you review this please? This MR is ready to be merged, only review is needed.

File: tensorflow/lite/delegates/xnnpack/README.md
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/README.md:345]: This is not dynamic range quantization. Models with dynamic range quantization quantized floating-point inputs to Conv2D/FC/etc, do quantized computations, then dequantize the outputs.

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:196]: Extra parenthesis not needed

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:247]: Compute per-channel initialization parameters at runtime.

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:275]: Initialize to 0, as this is the default value in `CreateQuantizationParameters`

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:293]: Combine with variable definition

**Review by Maratyszcza — CHANGES_REQUESTED** [review]
Thank you for the contribution, and apologies for the delays with the review. The PR needs some changes:
1. Compute per-channel quantization parameters at runtime rather than random-initialize.
2. Add support for the newly added TransposeConv operator

File: tensorflow/lite/delegates/xnnpack/README.md
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/README.md:345]: I am not quite sure about that. From my experience I have concluded that dynamic-range quantization implies that weights are quantized for storage, but at runtime almost all of them (not all because hybrid kernels are involved) are being dequantized and then all the computations are performed in floating-point. I got this understanding by inspecting converted TFLite models with Netron viewer. Some of my observations are described in [this](https://github.com/tensorflow/tensorflow/pull/48751#issue-867023983) message. Please correct me if I my conclusions are wrong.

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc]: Fixed

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc]: Fixed

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc]: Done

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc]: Fixed

**Comment by dev0x13** [comment]
@Maratyszcza Thank you for the review! All the changes are made.

File: tensorflow/lite/delegates/xnnpack/README.md
Comment (Maratyszcza) [reply tensorflow/lite/delegates/xnnpack/README.md:345]: TFLite converter creates operators with `FP32` inputs/outputs and `INT8` weights when using dynamic-range quantization. These operators are processed by so-called hybrid kernels, which dynamically quantize input and dequantize output. XNNPACK currently doesn't support hybrid schemas, so the models converted with dynamic range quantization will not on the XNNPACK delegate, and it would be confusing to state otherwise.

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:72]: Check that if channelwise weights are used, the static input has at least one dimension (i.e. isn't a scalar)

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:230]: Combine with resize

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/test_util.cc:85]: const size_t

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/test_util.cc:90]: Initialize min to `-std::numeric_limits<float>::infinity()` and max to `+std::numeric_limits<float>::infinity()` and remove these checks

**Review by Maratyszcza — CHANGES_REQUESTED** [review]
LG overall. Please revert the change in README and simplify the code per comments, and we're good to go.

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/test_util.cc]: Done

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/test_util.cc]: Fixed

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc]: Done

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:75]: Done

**Comment by dev0x13** [comment]
@Maratyszcza Thank you! All the suggested changes are made.

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (Maratyszcza) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:75]: Should be `if (INT8ChannelWiseWeights() && Input1Static())` ... + the same for input2

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (Maratyszcza) [inline tensorflow/lite/delegates/xnnpack/test_util.cc:89]: These lines can be further simplified using `std::min`, `std::max`

File: tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/binary_elementwise_tester.cc:78]: Sorry, my bad. Fixed

File: tensorflow/lite/delegates/xnnpack/test_util.cc
Comment (dev0x13) [reply tensorflow/lite/delegates/xnnpack/test_util.cc]: Done

**Review by Maratyszcza — APPROVED** [review]

**Comment by Maratyszcza** [comment]
LGTM. @gbaned could you merge?

**Comment by dev0x13** [comment]
I noticed that the PR was rolled back right after merge. Is there something wrong with it?

**Comment by Maratyszcza** [comment]
`transpose_conv_test` failed under sanitizers with `UndefinedBehaviorSanitizer: float-cast-overflow tensorflow/lite/delegates/xnnpack/test_util.cc:32:28`

**Comment by dev0x13** [comment]
@Maratyszcza Fixed this in [my fork](https://github.com/dev0x13/tensorflow/commit/bc81bf3c74c47bb07aea6c2d2dbb1887d83d71a9), but it seems that this PR cannot be reopened. What do I need to do in order to submit this fix to make my changes merged again? Thank you in advance!

**Comment by Maratyszcza** [comment]
`prelu_test` fails under asan too, with the same error

**Comment by dev0x13** [comment]
@Maratyszcza [Fixed](https://github.com/dev0x13/tensorflow/commit/efa0a06559b2f980fc967a9dde8338997aadb8ac)

**Comment by Maratyszcza** [comment]
Re-landed in a6d352f031f272972da0e6271de07e8b40cac74a

**Comment by dev0x13** [comment]
Thank you!

**Pull request merged**
